### PR TITLE
Feature/cdap 14022 configurable bootstrap

### PR DESCRIPTION
**Pull request by albertshau**
Added the bootstrap framework. This allows users to configure a bootstrap file that will be used to determine what steps need to be performed on a new CDAP instance. Replaces existing services for creating the native profile, default namespace, and loading system artifacts.

There are a lot of lines of code, but a lot of it is boilerplate. Most of the logic is in BootstrapService and BaseStepExecutor.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapConfig.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  public List<BootstrapStep> getSteps() {
+    // can be null when deserialized through gson
+    return steps == null ? Collections.emptyList() : Collections.unmodifiableList(steps);
+  }
+
+  /**
+   * Check that this is a valid config, throwing an exception if not.
+   *
+   * @throws IllegalArgumentException if the config is invalid
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapConfig.java:52]: method definition doesn't throw exception.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/FileBootstrapConfigProvider.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  @Override
+  public BootstrapConfig getConfig() {
+    try {
+      try (Reader reader = new FileReader(bootstrapFile)) {
+        return GSON.fromJson(reader, BootstrapConfig.class);
+      }
+    } catch (FileNotFoundException e) {
+      LOG.info("Bootstrap file {} does not exist. Default bootstrapping will be done.");
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/FileBootstrapConfigProvider.java:55]: missing `file` parameter in the log statements.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapService.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+                                     String.format("Unknown bootstrap step type '%s' for '%s'.",
+                                                   step.getType(), step.getLabel()));
+    }
+    return executor.execute(step.getLabel(), step.getArguments());
+  }
+
+  /**
+   * Reloads the bootstrap config.
+   */
+  public void reload() {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapService.java:169]: is this only used in test?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapService.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    } catch (IllegalArgumentException e) {
+      LOG.warn("Bootstrap step {} failed because it is malformed: {}", step.getLabel(), e.getMessage());
+      return new BootstrapStepResult(step.getLabel(), BootstrapStepResult.Status.FAILED, e.getMessage());
+    }
+
+    if (shouldSkip.test(step)) {
+      return new BootstrapStepResult(step.getLabel(), BootstrapStepResult.Status.SKIPPED, null);
+    }
+
+    BootstrapStepExecutor executor = executors.get(step.getType());
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapService.java]: can we rename the executors to be something else `bootstrapExecutorMap` maybe, it's confusing with `java.util.concurrent.Executors` since you also use ExecutorService in this class..

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapStep.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+package co.cask.cdap.internal.bootstrap;
+
+import com.google.gson.JsonObject;
+
+/**
+ * Definition for a single step in a bootstrap process. Defines what operation to perform, whether it should be executed
+ * every time CDAP starts up or just once, and any arguments required to perform the operation.
+ */
+public class BootstrapStep {
+  private final String label;
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapStep.java:27]: there is no requirement for label to be unique right ? just that the user can get unambiguous response if they follow unique labeling.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/BootstrapStepExecutor.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+  /**
+   * Execute the bootstrap step.
+   *
+   * @param label the step label
+   * @param arguments the arguments for the step
+   * @return the result of execution
+   * @throws InterruptedException if the step was interrupted
+   */
+  BootstrapStepResult execute(String label, JsonObject arguments) throws InterruptedException;
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/BootstrapStepExecutor.java]: since its a single object, might be better to name it `argument` or `argumentObject`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/BaseStepExecutor.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      return new BootstrapStepResult(label, BootstrapStepResult.Status.SUCCEEDED);
+    } catch (InterruptedException e) {
+      throw e;
+    } catch (Exception e) {
+      LOG.warn("Bootstrap step {} failed to execute", label, e);
+      return new BootstrapStepResult(label, BootstrapStepResult.Status.FAILED, e.getMessage());
+    }
+  }
+
+  private Class<T> getArgumentsType() {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/BaseStepExecutor.java]: will be good to add javadoc for this.

File: cdap-standalone/pom.xml [hunk 8]
```diff
@@ -633,6 +633,14 @@
...
                       </includes>
                     </resource>
+                    <!-- copy bootstrap file -->
+                    <resource>
+                      <directory>${project.parent.basedir}/cdap-common/src/main/resources/bootstrap</directory>
+                      <targetPath>bootstrap</targetPath>
+                      <includes>
+                        <include>sandbox.json</include>
+                      </includes>
+                    </resource>
```
Comment (gsps1) [inline cdap-standalone/pom.xml:643]: why is this required ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemArtifactLoader.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+  @Override
+  public void execute(EmptyArguments arguments) {
+    try {
+      artifactRepository.addSystemArtifacts();
+    } catch (Exception e) {
+      // retrying everything was the behavior before this logic got moved to the bootstrap framework.
+      // if there really is non-transient exception here, the retry time limit will eventually be hit
+      // and this step will be skipped.
+      throw new RetryableException(e);
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemArtifactLoader.java:43]: should we throw if its `InterruptedException` ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemProfileCreator.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    profileService.createIfNotExists(arguments.getId(), profile);
+  }
+
+  /**
+   * Arguments required to create a profile
+   */
+  static class Arguments extends ProfileCreateRequest implements Validatable {
+    private final String name;
+
+    Arguments(String name, String label, String description, ProvisionerInfo provisioner) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemProfileCreator.java:53]: it seems strange that default profile create request has label but not name. why is it so ?

**Review by gsps1 — COMMENTED** [review]
changes look good, just few comments/questions.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapConfig.java (hunk 1 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapConfig.java:52]: for runtime exceptions, we only document it but we don't put it in the method definition.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapService.java (hunk 3 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapService.java:169]: currently yes, but it will be called when we add the rest endpoint.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapStep.java (hunk 5 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapStep.java:27]: that's correct

File: cdap-standalone/pom.xml (hunk 8 above)
Comment (albertshau) [reply cdap-standalone/pom.xml:643]: this is needed for packaging purposes, to copy the bootstrap file to the right place. If you're asking why there's a separate file for distributed and sandbox, it's because distributed doesn't auto-start dataprep, but sandbox does. Will change the bootstrap files accordingly once the start program type is implemented

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemArtifactLoader.java (hunk 9 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemArtifactLoader.java:43]: loadSystemArtifacts() does not throw InterruptedException.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemProfileCreator.java (hunk 10 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemProfileCreator.java:53]: It's because the name comes from the path of the REST call, it's not actually a parameter in the request body.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapService.java (hunk 4 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapService.java]: done

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/BootstrapStepExecutor.java (hunk 6 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/BootstrapStepExecutor.java]: done

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/BaseStepExecutor.java (hunk 7 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/BaseStepExecutor.java]: done

**Comment by gsps1** [comment]
LGTM

**Review by gsps1 — APPROVED** [review]
